homeassistant/components/screenlogic/water_heater.py

```diff
@@ -0,0 +1,128 @@
+"""Support for a ScreenLogic Water Heater."""
+import logging
+
+from screenlogicpy.const import HEAT_MODE
+
+from homeassistant.components.water_heater import (
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+    WaterHeaterEntity,
+)
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS, TEMP_FAHRENHEIT
+
+from . import ScreenlogicEntity
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORTED_FEATURES = SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE
+
+HEAT_MODE_NAMES = HEAT_MODE.Names
+
+MODE_NAME_TO_MODE_NUM = {
+    HEAT_MODE_NAMES[num]: num for num in range(len(HEAT_MODE_NAMES))
+}
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up entry."""
+    entities = []
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator = data["coordinator"]
+
+    for body in data["devices"]["water_heater"]:
+        entities.append(ScreenLogicWaterHeater(coordinator, body))
+    async_add_entities(entities, True)
+
+
+class ScreenLogicWaterHeater(ScreenlogicEntity, WaterHeaterEntity):
+    """Represents the heating functions for a body of water."""
+
+    @property
+    def name(self) -> str:
+        """Name of the water heater."""
+        ent_name = self.body["heat_status"]["name"]
+        return f"{self.gateway_name} {ent_name}"
+
+    @property
+    def state(self) -> str:
+        """State of the water heater."""
+        return HEAT_MODE.GetFriendlyName(self.body["heat_status"]["value"])
+
+    @property
+    def min_temp(self) -> float:
+        """Minimum allowed temperature."""
+        return self.body["min_set_point"]["value"]
+
+    @property
+    def max_temp(self) -> float:
+        """Maximum allowed temperature."""
+        return self.body["max_set_point"]["value"]
+
+    @property
+    def current_temperature(self) -> float:
+        """Return water temperature."""
+        return self.body["last_temperature"]["value"]
+
+    @property
+    def target_temperature(self) -> float:
+        """Target temperature."""
+        return self.body["heat_set_point"]["value"]
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement."""
+        if self.config_data["is_celcius"]["value"] == 1:
+            return TEMP_CELSIUS
+        return TEMP_FAHRENHEIT
+
+    @property
+    def current_operation(self) -> str:
+        """Return operation."""
+        return HEAT_MODE_NAMES[self.body["heat_mode"]["value"]]
+
+    @property
+    def operation_list(self):
+        """All available operations."""
+        supported_heat_modes = [HEAT_MODE.OFF]
+        # Is solar listed as available equipment?
+        if self.coordinator.data["config"]["equipment_flags"] & 0x1:
+            supported_heat_modes.extend([HEAT_MODE.SOLAR, HEAT_MODE.SOLAR_PREFERED])
+        supported_heat_modes.append(HEAT_MODE.HEATER)
+
+        return [HEAT_MODE_NAMES[mode_num] for mode_num in supported_heat_modes]
```

Comment:
The only allowed modes are the states in the water heater entity integration.

https://developers.home-assistant.io/docs/core/entity/water-heater#states

Reply:
It might be a better fit to use the climate platform with heat only given the number of presets.

Reply:
Looking at the [Architecture repo](https://github.com/home-assistant/architecture/discussions/518), It appears that there is a desire to bring the `water_heater` platform up to par with `climate`. This seems to include adding "presets" functionality. Ideally, I had thought that I should have a device that heats water to be represented by a `water_heater` entity, also because there is no concept of a fan, but it seem like the `water_heater` platform is too new or maybe not yet mature enough, as several aquatic integrations are using the `climate` platform (iaqualink, smarttub). That said, the Pentair pool controllers that the Screenlogic system supports, in theory have the ability to support a cooling mode as well so `climate` may be better suited anyways.

So the questions I have are: 
1. Are we ok with permanently using the `climate` platform to represent a device that doesn't control the air temperature, in this case the _climate_ of a body of water (pool/spa)? I worry about being disruptive to the end user if at some point in the future we have to move back to `water_heater`.
2. If moving over to `climate`, how does this process work with the release of the integration? Will it be held until the temp controls are moved over to `climate`, or will the integration be released as-is and be a breaking change later?
3. Also, when addressing this (and the rest of the changes) in a new PR, should I keep and continue working in the current `dieselrabbit/screenlogic` branch, or delete this branch and create a new one? Still a bit of a git noob.

Thanks for your time.

Reply:
1. I think it's ok to use climate since we aren't strict with that categorization yet. Water heater would probably fit better but there's no time table for the improvements there.
2. Unless we revert this PR it will go out in the next minor release of HomeAssistant. If we don't want to release it, we could revert and then make a separate target branch based off dev branch and cherry pick this PR to that branch.That way we can continue the work in the separate target branch and make PRs to that.
3. Start in a fresh branch in the next PR based on latest target branch, eg dev branch.

Reply:
1. I'll start moving the `water-heater` over to `climate`.
2. Hopefully I'll get it done before the next release (unknown when that is).
3. Alright. Will delete this and start a new branch.

Reply:
https://calendar.google.com/calendar/u/0/r/week/2021/3/31?eid=NGRvOGNodGE0ZWg2cnM3M2pqMnUxdDZsOWdfMjAyMTAzMzEgcDA3bjk4Z28xMW9uYW1kMDhkMGttcTZqaHNAZw&ctz=America/Los_Angeles&pli=1&sf=true

For it to make the next release it needs to be merged before March 31st

Reply:
Moved over to climate platform. Basic on/off is in the mode, what type of heating is in the preset.